### ISSUE TITLE
Using Botex to implement LLM played experiment on Otree

### ISSUE THREAD

**Issue by Yizhi-91** [joachim-gassen/botex#16]
I am a new user of the Python package Botex.

I have come across several issues that prevent me from running the experiment. I emailed Professor Joachim Gassen and received many good advice. Here I will put a summary of these issues and feedback.

Firstly, the setup of BOTEX_DB in the 'secrets.env' file is not working for me. I used the following code to circumvent this issue:

botex.run_single_bot(
botex_db="botex.sqlite3",
url="http://localhost:8000/InitializeParticipant/zltqkbs5"
)

This code set up the default database for the GPT data at "botex.sqlite3".

Second, I wanted to implement multiple GPT-played agents together with some human players and assign the human players to certain roles in the Otree experiment.

The code to achieve this is:

sdict = botex.init_otree_session(
config_name = "Analysing_Political_Corruption",
npart = 7, # number of participants in the session, including bots and humans
is_human=[True, True, False, False, False, False, False],
botex_db = "botex.sqlite3",
otree_server_url = "http://localhost:8000/",
)

In this way, the first two players are human players and the last five players are GPT-played agents.

The new issue that I wanted to ask is related to the Otree-rest-key problem. I am using a third-party server in China called yanzhonglab.com. I don't know how to set up the rest-key. Is Heroku easier to use in this regard?

The error message is the following:

Request to "https://xu8uze.otree.yanzhonglab.com/api/sessions" failed with status code 403: HTTP Request Header otree-rest-key is incorrect

I found out that the rest-key is set up in the 'secrets.env' file and also in the settings file of the Otree folder. Now I have set both of them as '1234'. But the above error message still comes out. I don't know why this would happen.

Can you help me with this?

Many thanks!

**Comment by Yizhi-91** [comment]
I have finally found the API rest-key in yanzhonglab.com. There are no more questions up to now. I have successfully made the GPT to play five participants now.

**Comment by joachim-gassen** [comment]
Hi Yizhi: This is awesome news! Thank you for updating us on your progress. I will now close this issue for the time being but please feel free to post to it again if new questions arise. Good luck!

**Comment by joachim-gassen** [comment]
Closing for the time being....

**Comment by joachim-gassen** [comment]
Hi there:

This indicates a connection failure. As these can happen for multiple reasons (including temporary network failures) they are inherently hard to debug.

Can you reproduce the error? If not, then I would assume it is a temporary glitch. If yes, then please provide us with more info on the error. You mentioned chrome so did you experience a chrome crash prior to the error? Essentially I ask you to provide us full logs of the error including a description of what you did.

FWIW: I am developing on a MacBook Pro so I doubt that this per se is the culprit. But if the error reproduces can you also send us some more information on your hardware, just in case.

Thanks!

**Comment by Yizhi-91** [comment]
Hi Professor Gassen,

I tried to use another laptop (windows) computer to run the code. The error appears again. 

What I have succeeded is to use a desktop computer (windows) with the help of VPN to run the code with 10 players that are simulated by ChatGPT. 

However, when I wanted to implement more players simulated by GPT, the desktop computer is not enough (due to CPU issues). So I want to try using better computers. When I use laptop computers and VPN, I got the same error messages.

 Here is the first part of the error message:

INFO:httpx:HTTP Request: GET https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json "HTTP/1.1 200 OK"
Traceback (most recent call last):
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connectionpool.py", line 789, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connectionpool.py", line 490, in _make_request
    raise new_e
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connectionpool.py", line 466, in _make_request
    self._validate_conn(conn)
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connectionpool.py", line 1095, in _validate_conn
    conn.connect()
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connection.py", line 730, in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\connection.py", line 909, in _ssl_wrap_socket_and_match_hostname
    ssl_sock = ssl_wrap_socket(
               ^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\util\ssl_.py", line 469, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\OneDrive\Otree\Otree\venv\Lib\site-packages\urllib3\util\ssl_.py", line 513, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\AppData\Local\Programs\Python\Python312\Lib\ssl.py", line 455, in wrap_socket
    return self.sslsocket_class._create(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wangy\AppData\Local\Programs\Python\Python312\Lib\ssl.py", line 1042, in _create
    self.do_handshake()
  File "C:\Users\wangy\AppData\Local\Programs\Python\Python312\Lib\ssl.py", line 1320, in do_handshake
    self._sslobj.do_handshake()
ConnectionResetError: [WinError 10054] 远程主机强迫关闭了一个现有的连接(the remote computer shuts down an existing link)。

Can you help me with this? 

Many thanks,
Y.

**Comment by joachim-gassen** [comment]
Sorry but this error happens outside of the botex code so I cannot offer any help here. Some general thoughts that might help to get you started:


1. When filing a potential botex error, please always include the botex log. You can enable logging in your python code by including `import logging; logging.basicConfig(level=logging.INFO)` in your code before calling any botex code.
2. It seems that you are using a VPN to access oTree resources. When using botex and oTree for experiments, we tend to run both oTree and botex locally as this rules out network glitches as a potential source of error. 
3. When running experiment with multiple participants, running the resulting bots in parallel can put excessive strain on your computing environment. Each botex bot has its own thread and uses its own chrome instance to access the oTree webpages. Also, concurrent bots increase the risk that you hit usage limits of your commercial LLM API. So, to run large experiments with multiple participants, we normally split them up in multiple sessions with up to 10 participants each and run the resulting sessions sequentially. This way you can run experiments with any amount of participants.

Good luck!

Joachim

**Comment by Yizhi-91** [comment]
Hi Joachim,

I have found out the problem now. It is the setting of my Astrill VPN. For users in China, you should use the option WireGuard so that Chrome can get access to GPT. 

Thank you!
Y.

**Comment by joachim-gassen** [comment]
Great to hear that it works now. Good luck with the next steps!
